pkg/apis/ceph.rook.io/v1/webhook.go

```diff
@@ -0,0 +1,89 @@
+package v1
+
+import (
+	"github.com/coreos/pkg/capnslog"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"strconv"
+)
+
+var (
+	webhookName = "rook-ceph-webhook"
+	logger      = capnslog.NewPackageLogger("github.com/rook/rook", webhookName)
+)
+
+// compile-time assertions ensures CephCluster implements webhook.Validator so a webhook builder
+// will be registered for the validating webhook.
+var _ webhook.Validator = &CephCluster{}
+
+func (in *CephCluster) ValidateCreate() error {
+	logger.Info("validate create cephcluster")
+	//If external mode enabled, then check if other fields are empty
+	if in.Spec.External.Enable {
+		if in.Spec.Mon != (MonSpec{}) || in.Spec.Dashboard != (DashboardSpec{}) || in.Spec.Monitoring != (MonitoringSpec{}) || in.Spec.DisruptionManagement != (DisruptionManagementSpec{}) || len(in.Spec.Mgr.Modules) > 0 || len(in.Spec.Network.Provider) > 0 || len(in.Spec.Network.Selectors) > 0 {
+			return errors.New("invalid create : external mode enabled cannot have mon,dashboard,monitoring,network,disruptionManagement,storage fields in CR")
+		}
+	}
+	return nil
+}
+
+func (in *CephCluster) ValidateUpdate(old runtime.Object) error {
+	logger.Info("validate update cephcluster")
+	occ := old.(*CephCluster)
+	return validateUpdatedCephCluster(in, occ)
+}
+
+func (in *CephCluster) ValidateDelete() error {
+	return nil
+}
+
+func validateUpdatedCephCluster(updatedCephCluster *CephCluster, found *CephCluster) error {
+	if updatedCephCluster.Spec.DataDirHostPath != found.Spec.DataDirHostPath {
+		return errors.Errorf("invalid update: DataDirHostPath change from %q to %q is not allowed", found.Spec.DataDirHostPath, updatedCephCluster.Spec.DataDirHostPath)
+	}
+
+	if updatedCephCluster.Spec.Network.HostNetwork != found.Spec.Network.HostNetwork {
+		return errors.Errorf("invalid update: HostNetwork change from %q to %q is not allowed", strconv.FormatBool(found.Spec.Network.HostNetwork), strconv.FormatBool(updatedCephCluster.Spec.Network.HostNetwork))
+	}
+
+	if updatedCephCluster.Spec.Network.Provider != found.Spec.Network.Provider {
+		return errors.Errorf("invalid update: Provider change from %q to %q is not allowed", found.Spec.Network.Provider, updatedCephCluster.Spec.Network.Provider)
+	}
+
+	return nil
+}
+
+var _ webhook.Validator = &CephBlockPool{}
+
+func (in *CephBlockPool) ValidateCreate() error {
+	logger.Infof("validate create cephblockpool %v", in)
```

Comment:
We have pools embedded in other CRs such as the CephObjectStore and CephFilesystem. The same validations would apply to those CRs as well, so these pool checks could be helpers that other CR validations also call for a pool.

Reply:
Added helper function ValidatePoolSpec